File: Design-Patterns-Using-Java/State/StatePattern.java

class BulbControl {
    private BulbState current;

    BulbControl() {
        this.current = new Off();
    }

    void setState(BulbState state) {
        this.current = state;
    }

    void flip() {
        this.current.flip(this);
    }

    String toStringState() {
        return this.current.toString();
    }
}

interface BulbState {
    void flip(BulbControl bc);
    String toString();
}

class On implements BulbState {
    @Override
    public void flip(BulbControl bc) {
        bc.setState(new Off());
    }

    @Override
    public String toString() {
        return "On";
    }
}

class Off implements BulbState {
    @Override
    public void flip(BulbControl bc) {
        bc.setState(new On());
    }

    @Override
    public String toString() {
        return "Off";
    }
}

// Client code.
public class StatePattern {
    public static void main(String[] args) {
        BulbControl c = new BulbControl();
        c.flip();
        System.out.println(c.toStringState());
        c.flip();
        System.out.println(c.toStringState());
    }
}

/* 
On
Off
*/

File: Design-Patterns-Using-CSharp/State/StatePattern.cs
using System;

class BulbControl
{
    private IBulbState current;

    public BulbControl()
    {
        this.current = new Off();
    }

    public void SetState(IBulbState state)
    {
        this.current = state;
    }

    public void Flip()
    {
        this.current.Flip(this);
    }

    public string ToStringState()
    {
        return this.current.ToString();
    }
}

interface IBulbState
{
    void Flip(BulbControl bc);
    string ToString();
}

class On : IBulbState
{
    public void Flip(BulbControl bc)
    {
        bc.SetState(new Off());
    }

    public override string ToString()
    {
        return "On";
    }
}

class Off : IBulbState
{
    public void Flip(BulbControl bc)
    {
        bc.SetState(new On());
    }

    public override string ToString()
    {
        return "Off";
    }
}

// Client code.
public class StatePattern
{
    public static void Main(string[] args)
    {
        BulbControl c = new BulbControl();
        c.Flip();
        Console.WriteLine(c.ToStringState());
        c.Flip();
        Console.WriteLine(c.ToStringState());
    }
}
